Fixes error propagation and resizes during MFA

ibeckermayer opened the pull request:
Adds mechanisms for withholding resizes

1. A new mechanism is added to withhold all non-MFA TDP messages during the MFA ceremony within lib/web/desktop.go.
2. A new mechanism is added to withhold all resizes when the rdpclient/client.go is not yet ready to receive messages. (Non-resize messages are still just dropped).

This fixes https://github.com/gravitational/teleport/issues/41515

Also fixes the error propagation issue evident in https://github.com/gravitational/teleport/issues/41515, see point 1 in https://github.com/gravitational/teleport/issues/41515#issuecomment-2110326395

changelog: Solved a bug that was causing Windows Desktop sessions to fail when the screen was resized during an MFA ceremony

Commit:
Update 'failed' statusText only if a previous error has not already been displayed to the user. This way the first error message (which is usually the most informative as to what's gone awry) is displayed to the user in the case of an error cascade. (E.g. tdp error is sent back, the server to closes the websocket, and that close causes a client websocket error. In that case, we want to display the tdp error that was sent back, not the websocket is closed error.)

Commit:
Adds mechanisms for withholding resizes

1. A new mechanism is added to withhold all non-MFA TDP
   messages during the MFA ceremony within lib/web/desktop.go
2. A new mechanism is added to withhold all resizes when the
   rdpclient/client.go is not yet ready to receive messages.
   (Non-resize messages are still just dropped).

This fixes https://github.com/gravitational/teleport/issues/41515

zmb3 commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -365,7 +368,7 @@ func (c *Client) startInputStreaming(stopCh chan struct{}) error {
 
 	// Remember mouse coordinates to send them with all CGOPointer events.
-	var mouseX, mouseY uint32
+	var withheldResize *tdp.ClientScreenSpec
```
The comment about remembering mouse coordinates no longer seems accurate here.

ibeckermayer replied:
f36db5e5f1c7dcaa346b1438557a8b6005c5ba35

zmb3 commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -385,284 +388,310 @@ func (c *Client) startInputStreaming(stopCh chan struct{}) error {
+	}
+}
+
+func (c *Client) handleMessage(msg tdp.Message) error {
```
`handleMessage` seems too vague of a name - is it an incoming message or an outgoing message?

ibeckermayer replied:
f36db5e5f1c7dcaa346b1438557a8b6005c5ba35

zmb3 commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -385,284 +388,310 @@ func (c *Client) startInputStreaming(stopCh chan struct{}) error {
+			return trace.Errorf("RDPResponsePDU failed: %v", errCode)
 		}
+	default:
+		c.cfg.Logger.WarnContext(context.Background(),
+			fmt.Sprintf("Skipping unimplemented TDP message type %T", msg))
```
We no longer need to use Sprintf here. Forrest made a `TypeAttr` utility:
```suggestion
		c.cfg.Logger.WarnContext(context.Background(), "Skipping unimplemented TDP message", "type", logutils.TypeAttr(msg)))
```

zmb3 commented on lib/web/desktop.go:
```diff
@@ -218,7 +226,20 @@ const (
 )
 
-func (h *Handler) desktopTLSConfig(ctx context.Context, ws *websocket.Conn, clusterClient auth.ClientI, sessCtx *SessionContext, desktopName, username, siteName string) (_ *tls.Config, err error) {
+// desktopTLSConfig returns a TLS config for connecting to a Windows Desktop Service.
+// It also performs MFA if required.
+//
+// This function returns the TLS config, and a list of withheld TDP messages if MFA was
+// required and any were received before the MFA message was sent. It is the caller's
+// responsibility to send these withheld messages through to the server after the MFA
+// ceremony is completed.
```
I think the long confusing comment here is an indicator that `desktopTLSConfig` has become unwieldy. It started out as a way to build a `tls.Config` for the connection, now it withholds messages, does MFA, etc.

At a minimum we should rename it, but I wonder if we can come up with a better design so that we don't have to plumb these withheld messages through.

ibeckermayer replied:
I made `withheldMessages` a member of `Handler` and restructured things to make the steps more readily apparent from the function names: 6aa0331cda8fa820514dc896c6e75a7f56f45b30

zmb3 reviewed (commented):
Feels like this will work fine but I wonder if there's a better way.

Also, it would be great to start to get some test coverage in here, especially as we fix bugs like this one.

Commit:
Remove superfluous comment, rename handleMessage to handleTDPInput

Commit:
Refactors mfa and desktop handlers
to use a single function to handle checking if MFA is required,
and refactors the desktop handler to make the TLS configuration
creation more clear.

Commit:
Merge branch 'master' into isaiah/fix-error-propagation-and-resize-mfa

Commit:
use idiomatic TypeAttr

zmb3 reviewed (approved):
Looks good. Please make sure to test with and without MFA required, and with a resizing desktop as well as a fixed-size one.

zmb3 commented:
Let's try to get the backport merged by EOD Monday so it makes it into the Tuesday release.

Commit:
Only log a warning when we try to send TDP message on a closed ws

The only time I know this happens is when the user tries to resize on when
an "Active Session" warning modal is shown, in which case the session was
failing. In reality we can just ignore this scenario. If the websocket isn't
yet open, we don't want to send messages. If it closes, then the onclose
handler will take care of cleaning up the session, we don't need to error
if any straggling messages are attempted to be sent.

Commit:
Merge branch 'master' into isaiah/fix-error-propagation-and-resize-mfa
